modules/nw-operator-vsphere-multiple-nic.adoc

```diff
@@ -0,0 +1,75 @@
+
+//Bare metal only?
+[IMPORTANT]
+====
+Each NIC must have access to link:https://quay.io/repository/[Quay.io] to obtain packages that are required for the installation of your cluster.
+====
+
+If you do not configure internal and external CIDR implementations in the {vmw-short} CCM configuration, the {vmw-short} CCM selects the first NIC that the host operating system reports as the primary network interface. This configuration can cause new nodes that associate with a `MachineSet` object with a single NIC to become unusable as each new node receives the `node.cloudprovider.kubernetes.io/uninitialized` taint. In certain situations, the reported network interface might not be the expected network interface. These situations can cause communication issues with the Kubernetes API server that can cause installation of the cluster to fail.
+
+[IMPORTANT]
+====
+If you configured multiple NICs for a network implementation but intend to use only a single NIC for your current needs, ensure that you specify the same IP address for the internal and external CIDR implementations. This configuration ensures that the CCM selects your preferred NIC and not the first NIC that the host operating system reports.
+====
+
+.Example error message because of the {vmw-short} CCM selecting an incompatible NIC
+[source, text]
+----
+ERROR Bootstrap failed to complete: timed out waiting for the condition
+ERROR Failed to wait for bootstrapping to complete. This error usually happens when there is a problem with control plane hosts that prevents the control plane operators from creating the control plane.
+----
+
+.Prerequisites
+
+* You created Kubernetes manifest files for your {product-title} cluster.
+
+.Procedure
+
+. From the directory where you store your {product-title} cluster manifest files, open the `manifests/cluster-infrastructure-02-config.yml` manifest file.
+
+. Add a `nodeNetworking` object to the file and specify internal and external network subnet CIDR implementations for the object.
++
+.Example of a configured `cluster-infrastructure-02-config.yml` manifest file
+[source,yaml]
+----
+apiVersion: config.openshift.io/v1
+kind: Infrastructure
+metadata:
+  name: cluster
+spec:
+  cloudConfig:
+    key: config
+    name: cloud-provider-config
+  platformSpec:
+    type: VSphere
+    vsphere:
+      failureDomains:
+      - name: generated-failure-domain
+      ...
+      nodeNetworking:
+        external: <1>
+          networkSubnetCidr: 192.168.100.0/24
```

Comment:
This a `[]string`?  It needs to be

```
  - 192.168.100.0/24
 
```

It looks like the other examples are wrong.  We need to double-check.

Reply:
https://github.com/openshift/origin/blob/release-4.16/vendor/github.com/openshift/api/config/v1/types_infrastructure.go#L1305

```go
type VSpherePlatformNodeNetworkingSpec struct {
	// networkSubnetCidr IP address on VirtualMachine's network interfaces included in the fields' CIDRs
	// that will be used in respective status.addresses fields.
	// ---
	// + Validation is applied via a patch, we validate the format as cidr
	// +listType=set
	// +optional
	NetworkSubnetCIDR []string `json:"networkSubnetCidr,omitempty"
```

Reply:
This seems to do patch the values.

```
oc patch infrastructure cluster --type='json' -p='[{"op": "add", "path": "/spec/platformSpec/vsphere/nodeNetworking/internal/networkSubnetCidr", "value": ["'$(oc get infrastructures cluster -o jsonpath='{.status.platformStatus.vsphere.machineNetworks[0]}')'"]}, {"op": "add", "path": "/spec/platformSpec/vsphere/nodeNetworking/external/networkSubnetCidr", "value": ["'$(oc get infrastructures cluster -o jsonpath='{.status.platformStatus.vsphere.machineNetworks[0]}')'"]}]'
```

Reply:
Probably best to use user-replaced values, such as:

![Screenshot from 2024-06-25 14-11-35](https://github.com/openshift/openshift-docs/assets/57954076/6d433b7e-0c21-46f2-bf46-431d82cb1b3d)

WDYT?

Reply:
I guess since this is 4.16 we can re-use the `machineNetworks` field already in the spec.  `machineNetworks` is populated by the installer in 4.16

```yaml
  spec:
    platformSpec:
      type: VSphere
      vsphere:
        apiServerInternalIPs:
        - 10.38.247.2
        - fd65:a1a8:60ad:1237::4
        ingressIPs:
        - 10.38.247.3
        - fd65:a1a8:60ad:1237::5
        machineNetworks:
        - 10.38.247.0/25
        - fd65:a1a8:60ad:1237::/64
        nodeNetworking:
          external: {}
          internal: {}
 ```

So proper ref would be `spec.platformSpec.vsphere.machineNetworks` as in `oc get  infrastructures cluster -o jsonpath='{.spec.platformSpec.vsphere.machineNetworks}'`


```
        nodeNetworking:
          external:
            networkSubnetCidr:
            - <machine_network_cidr_ipv4>
            - <machine_network_cidr_ipv6>
          internal:
            networkSubnetCidr:
            - <machine_network_cidr_ipv4>
            - <machine_network_cidr_ipv6>
```

It looks like I tested the dualstack syntax back 4.14.

We really should future-proof the docs with IPv6 examples.

Reply:
Err nevermind, this is pre-install, so `spec.platformSpec.vsphere.machineNetworks` is not populated yet.

We could refer to the `.networking.machineNetwork` field in the `install-confg.yaml`

Reply:
Thanks. I updated to state:

![Screenshot from 2024-06-26 12-03-09](https://github.com/openshift/openshift-docs/assets/57954076/fde4dc7d-42b8-479c-a5ad-686ab2517ff6)

Maybe I could remove from the YAML as it could be assumed for a vSphere configuration?

```
     vcenters:
      - datacenters:
        - sample-dc
        port: 443
        server: vcenter.example.com
status:
  ...
  platformStatus:
    type: VSphere
```

Reply:
+1 to omitting that. The original KCS did too. An ellipse to show there's more content expected should take care of it and keeps the focus on the thing we're modifying here.

Reply:
Thanks. Removed this section.